src/CoreShop/Bundle/ResourceBundle/Pimcore/CacheResourceMarshaller.php

```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * CoreShop.
+ *
+ * This source file is subject to the GNU General Public License version 3 (GPLv3)
+ * For the full copyright and license information, please view the LICENSE.md and gpl-3.0.txt
+ * files that are distributed with this source code.
+ *
+ * @copyright  Copyright (c) CoreShop GmbH (https://www.coreshop.org)
+ * @license    https://www.coreshop.org/license     GNU General Public License version 3 (GPLv3)
+ */
+
+declare(strict_types=1);
+
+namespace CoreShop\Bundle\ResourceBundle\Pimcore;
+
+use CoreShop\Component\Resource\Model\ResourceInterface;
+use CoreShop\Component\Resource\Pimcore\Model\PimcoreModelInterface;
+use Doctrine\ORM\EntityManagerInterface;
+use Pimcore\Model\DataObject\ClassDefinition\Data;
+use Pimcore\Model\DataObject\Concrete;
+use Symfony\Component\Cache\Marshaller\DefaultMarshaller;
+use Symfony\Component\Cache\Marshaller\MarshallerInterface;
+
+class CacheResourceMarshaller implements MarshallerInterface
+{
+    protected MarshallerInterface $defaultMarshaller;
+    protected EntityManagerInterface $entityManager;
+
+    public function __construct(EntityManagerInterface $entityManager, MarshallerInterface $defaultMarshaller = null)
+    {
+        $this->defaultMarshaller = $defaultMarshaller ?? new DefaultMarshaller();
+        $this->entityManager = $entityManager;
+    }
+
+    public function marshall(array $values, ?array &$failed): array
+    {
+        foreach ($values as $data) {
+            if ($data instanceof Concrete) {
+                $class = $data->getClass();
+
+                if (!$class) {
+                    continue;
+                }
+
+                /**
+                 * @var Data&CacheMarshallerInterface $fd
+                 * @psalm-var Data&CacheMarshallerInterface $fd
+                 */
+                foreach ($class->getFieldDefinitions() as $fd) {
+                    if (!$fd instanceof CacheMarshallerInterface) {
+                        continue;
+                    }
+
+                    if (!$fd instanceof Data) {
+                        continue;
+                    }
+
+                    $data->setObjectVar(
+                        $fd->getName(),
+                        $fd->marshalForCache($data, $data->getObjectVar($fd->getName()))
+                    );
```

Comment:
`$data` is the already deep-copied version of the original object, right? Otherwise, we'd override the properties and the original object wouldn't be usable anymore.

Reply:
There is no deep-copy for cache. The idea here is to marshal the data per type and store it into the Object as anything you want. This only works as long as the properties are not fully typed. In my test case, I replace the `TaxRule` with the `id` of the `TaxRule`. The unmarshal then does the reverse. 

It is basically the same as the DeepCopy marshaller things work, but without the heavy overhead of DeepCopy.